torch/distributed/fsdp/_flat_param.py

```diff
@@ -2707,15 +2707,15 @@ def _warn_use_fake_reduce(log: logging.Logger, warning: str):
 
 @torch._dynamo.allow_in_graph
 def _same_storage(a, b):
-    return a._typed_storage()._data_ptr() == b._typed_storage().data_ptr()
+    return a.untyped_storage().data_ptr() == b.untyped_storage().data_ptr()
 
 
 @torch._dynamo.allow_in_graph
 def _same_storage_size(a: torch.Tensor, b: int):
-    return a._typed_storage()._size() == b
+    return a.untyped_storage().size() == b
```

Comment:
I think that this is actually not correct because `_typed_storage()._size()` returns a value that is `# of bytes per element` times larger than `untyped_storage().size()` (which assumes 1-byte elements).

I am guessing that the `_same_storage_size()` check is failing now when it should not. Could you try something like the following:
```suggestion
    return a.untyped_storage().size() // a.element_size() == b
```